org.eclipse.jgit/src/org/eclipse/jgit/internal/storage/dfs/DfsBlockCaffeineCache.java

```diff
@@ -0,0 +1,244 @@
+
+    /** Reverse index from DfsPackKey to the DfsPackDescription. */
+    private final Map<DfsPackKey, DfsPackDescription> reversePackDescriptionIndex;
+
+    /** Cache of Dfs blocks and Indices. */
+    private final Cache<DfsPackKeyWithPosition, Ref> dfsBlockAndIndicesCache;
+
+    public static void reconfigure(final DfsBlockCaffeineCacheConfig cacheConfig) {
+        DfsBlockCache.setInstance(new DfsBlockCaffeineCache(cacheConfig));
+    }
+
+    private DfsBlockCaffeineCache(DfsBlockCaffeineCacheConfig cacheConfig) {
+        maxStreamThroughCache = (long) (cacheConfig.getCacheMaximumSize() * cacheConfig.getStreamRatio());
+
+        blockSize = cacheConfig.getBlockSize();
+
+        packFileCache = new ConcurrentHashMap<>(16, 0.75f, 1);
+        reversePackDescriptionIndex = new ConcurrentHashMap<>(16, 0.75f, 1);
+
+        dfsBlockAndIndicesCache = Caffeine.newBuilder()
+                .removalListener(this::cleanUpIndices)
+                .maximumWeight(cacheConfig.getCacheMaximumSize())
+                // the estimated retained bytes is estimated based on the fields each object contains
+                // plus the key/entry reference itself
+                // DfsPackKey: {
+                //   final int hash;                4 bytes
+                //   final AtomicLong cachedSize;   28 bytes
+                // }
+                // final long position;             8 bytes
+                // final int size;                  4 bytes
+                // key, value reference             8 * 2 bytes
+                .weigher((DfsPackKeyWithPosition keyWithPosition, Ref ref) -> ref == null? 60 : 60 + ref.getSize())
+                .recordStats()
+                .build();
+    }
+
+    private void cleanUpIndices(DfsPackKeyWithPosition keyWithPosition, Ref ref, RemovalCause cause) {
+        if (keyWithPosition != null && ref != null) {
+            DfsPackKey key = keyWithPosition.getDfsPackKey();
+            long position = keyWithPosition.getPosition();
+
+            if (position < 0) {
+                // if it's an index, remove the whole packFile
+                cleanUpIndicesIfExists(key);
+            } else {
+                // if it's not an index, decrease the cached size
+                // remove the whole packFile if cachedSize is below 0
+                if (key.cachedSize.addAndGet(-ref.getSize()) <= 0) {
+                    cleanUpIndicesIfExists(key);
+                }
+            }
+        }
+    }
+
+    private void cleanUpIndicesIfExists(DfsPackKey key) {
+        if (key != null) {
+            DfsPackDescription description = reversePackDescriptionIndex.remove(key);
+            if (description != null) {
+                packFileCache.remove(description);
+            }
+        }
+    }
+
+    boolean shouldCopyThroughCache(long length) {
+        return length <= maxStreamThroughCache;
+    }
+
+    DfsPackFile getOrCreate(DfsPackDescription description, DfsPackKey key) {
```

Comment:
I feel like there's an edge case where I could just call getOrCreate(...) a bunch of times and never actually load anything into the cache, and the two maps would never get cleared out